File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def traverse(start):
            if start == 0: 
                dp[start] =  nums[0]
                return nums[0]
            if start == 1:
                dp[start] = nums[1]
                return nums[1]
            if dp[start] != -1 : return dp[start]
            li = []
            for i in range(start-2 , -1  , -1):
                li.append(traverse(i))
            dp[start] = max(li) + nums[start]
            return dp[start]

        dp = [-1] * len(nums)
        for i in range(len(nums)-1 , -1 , -1):
            traverse(i)
        return max(dp)

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int traverse(int indi  , vector<int> & nums , vector<int> &dp){

        if (indi <0 ) return 0;
        if (dp[indi] != -1) return dp[indi];
        int nottake = traverse(indi - 1 , nums , dp);
        int take = nums[indi] + traverse(indi - 2 , nums, dp);

        return dp[indi] = max(nottake , take);
    }
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size()  , -1);
        return traverse(nums.size()-1 , nums , dp);
    }
};